server/app/data_harvester.py

```diff
@@ -0,0 +1,187 @@
+
+    To resolve these issues, this class will:
+    (1) Scan for new data uploaded to S3
+    (2) Download the data, translate the skeleton to the standard skeleton, and re-upload it to S3 under a protected partition
+    """
+    bucket: str
+    deployment: str
+    target_folder: str
+    target_skeleton: str
+    index: ReactiveS3Index
+    queue: List[SubjectSnapshot]
+
+    def __init__(self, bucket: str, deployment: str, target_folder: str, target_skeleton: str, disable_pubsub: bool) -> None:
+        self.bucket = bucket
+        self.deployment = deployment
+        self.target_folder = target_folder
+        self.target_skeleton = target_skeleton
+        self.queue = []
+        self.index = ReactiveS3Index(bucket, deployment, disable_pubsub)
+        self.index.addChangeListener(self.onChange)
+        self.index.refreshIndex()
+        if not disable_pubsub:
+            self.index.registerPubSub()
+
+    def onChange(self):
+        print('S3 CHANGED!')
+
+        # 1. Collect all Trials
+        new_queue: List[SubjectSnapshot] = []
+        for folder in self.index.listAllFolders():
+            if self.index.hasChildren(folder, ['trials/', '_subject.json']):
+                if not folder.endswith('/'):
+                    folder += '/'
+                subject = SubjectSnapshot(self.index, folder, self.target_folder, self.target_skeleton)
+                if subject.ready_to_snapshot():
+                    print('Ready to snapshot: ' + str(subject.path))
+                    new_queue.append(subject)
+
+        print('Updating queue to have ' + str(len(new_queue)) + ' items')
+        self.queue = new_queue
+
+    def processQueueForever(self):
+        """
+        This busy-waits on the queue updating, and will process the head of the queue one at a time when it becomes available.
+
+        While processing, this blocks, so even though the queue is updating in the background, that shouldn't change the outcome of this process.
+        """
+        while True:
+            try:
+                if len(self.queue) > 0:
+                    print('Processing queue: ' + str(len(self.queue)) + ' items remaining')
+                    self.queue[0].copy_snapshot()
+                    self.queue.pop(0)
+            except Exception as e:
+                print(e)
+            time.sleep(1)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description='Run a data harvesting daemon.')
+    parser.add_argument('target_folder', type=str, default='rajagopal_with_arms',
+                        help='The folder where the converted data will be stored')
+    parser.add_argument('target_skeleton', type=str, default='data/PresetSkeletons/Rajagopal2015_CMUMarkerSet.osim',
+                        help='The OSIM skeleton to convert all the data to')
+    parser.add_argument('--bucket', type=str, default="biomechanics-uploads161949-dev",
+                        help='The S3 bucket to access user data in')
```

Comment:
Did you mean to keep the default values for the three arguments above?

Reply:
Fixed